pytorch_lightning/core/step_result.py

```diff
@@ -0,0 +1,327 @@
+from typing import Optional, Dict, Union
+from torch import Tensor
+import torch
+from copy import copy
+
+
+class Result(Dict):
```

Comment:
Inheriting from dict is going to bite us... 
If you inherit from dict you would expect that update, pop etc. work... but they don't!!!
It's too dangerous. Please let's be careful here. 
Why do we need to mimic a dictionary?

Reply:
We can use MutableMapping

Reply:
we shall also revisit
 https://github.com/PyTorchLightning/pytorch-lightning/blob/d2967d9305b42c9260f821f2b7fb43fbf19ca1aa/pytorch_lightning/utilities/parsing.py#L96
 in separate PR?